koschei/frontend/api.py

```diff
@@ -0,0 +1,52 @@
+# Copyright (C) 2017 Red Hat, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Author: Mikolaj Izdebski <mizdebsk@redhat.com>
+
+from flask import jsonify, request
+from sqlalchemy.orm import joinedload
+from koschei.frontend import app, db
+from koschei.models import Package, Collection
+
+
+def build_to_json(build):
+    if build:
+        return {'task_id': build.task_id}
+
+
+def package_to_json(package):
+    if package:
+        return {'name': package.name,
+                  'collection': package.collection.name,
+                  'state': package.state_string,
+                  'last_complete_build': build_to_json(package.last_complete_build)}
+
+
+def packages_to_json(packages):
+    return [package_to_json(p) for p in packages]
+
+
+@app.route('/api/v1/packages')
+def list_packages():
+    query = db.query(Package).join(Collection)\
+        .options(joinedload(Package.collection))\
+        .options(joinedload(Package.last_complete_build))
+    if 'name' in request.args:
+        query = query.filter(Package.name.in_(request.args.getlist('name')))
+    if 'collection' in request.args:
+        query = query.filter(Collection.name.in_(request.args.getlist('collection')))
+    packages = query.order_by(Package.name).all()
```

Comment:
For me this query takes less than one millisecond. Tested on full production DB dump:
```
koschei=# EXPLAIN ANALYZE SELECT package.id AS package_id, package.base_id AS package_base_id, package.name AS package_name, package.collection_id AS package_collection_id, package.arch_override AS package_arch_override, package.skip_resolution AS package_skip_resolution, package.last_complete_build_id AS package_last_complete_build_id, package.last_complete_build_state AS package_last_complete_build_state, package.last_build_id AS package_last_build_id, package.resolved AS package_resolved, package.static_priority AS package_static_priority, package.manual_priority AS package_manual_priority, package.build_priority AS package_build_priority, package.dependency_priority AS package_dependency_priority, package.tracked AS package_tracked, package.blocked AS package_blocked, package.scheduler_skip_reason AS package_scheduler_skip_reason, build_1.id AS build_1_id, build_1.package_id AS build_1_package_id, build_1.state AS build_1_state, build_1.task_id AS build_1_task_id, build_1.started AS build_1_started, build_1.finished AS build_1_finished, build_1.epoch AS build_1_epoch, build_1.version AS build_1_version, build_1.release AS build_1_release, build_1.repo_id AS build_1_repo_id, build_1.cancel_requested AS build_1_cancel_requested, build_1.deps_resolved AS build_1_deps_resolved, build_1.real AS build_1_real, collection_1.id AS collection_1_id, collection_1."order" AS collection_1_order, collection_1.name AS collection_1_name, collection_1.display_name AS collection_1_display_name, collection_1.secondary_mode AS collection_1_secondary_mode, collection_1.target AS collection_1_target, collection_1.dest_tag AS collection_1_dest_tag, collection_1.build_tag AS collection_1_build_tag, collection_1.branch AS collection_1_branch, collection_1.bugzilla_product AS collection_1_bugzilla_product, collection_1.bugzilla_version AS collection_1_bugzilla_version, collection_1.priority_coefficient AS collection_1_priority_coefficient, collection_1.build_group AS collection_1_build_group, collection_1.latest_repo_id AS collection_1_latest_repo_id, collection_1.latest_repo_resolved AS collection_1_latest_repo_resolved, collection_1.poll_untracked AS collection_1_poll_untracked FROM package JOIN collection ON collection.id = package.collection_id LEFT OUTER JOIN build AS build_1 ON build_1.id = package.last_complete_build_id LEFT OUTER JOIN collection AS collection_1 ON collection_1.id = package.collection_id WHERE package.name IN ('rnv', 'fop') AND collection.name IN ('f26') ORDER BY package.name;
                                                                   QUERY PLAN                                                                   
------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=46.76..46.77 rows=2 width=435) (actual time=0.132..0.133 rows=2 loops=1)
   Sort Key: package.name
   Sort Method: quicksort  Memory: 25kB
   ->  Nested Loop Left Join  (cost=9.29..46.75 rows=2 width=435) (actual time=0.079..0.109 rows=2 loops=1)
         Join Filter: (collection_1.id = package.collection_id)
         Rows Removed by Join Filter: 4
         ->  Nested Loop Left Join  (cost=9.29..45.63 rows=2 width=124) (actual time=0.068..0.089 rows=2 loops=1)
               ->  Nested Loop  (cost=8.86..28.72 rows=2 width=69) (actual time=0.058..0.069 rows=2 loops=1)
                     Join Filter: (package.collection_id = collection.id)
                     Rows Removed by Join Filter: 2
                     ->  Seq Scan on collection  (cost=0.00..1.04 rows=1 width=4) (actual time=0.009..0.010 rows=1 loops=1)
                           Filter: ((name)::text = 'f26'::text)
                           Rows Removed by Filter: 2
                     ->  Bitmap Heap Scan on package  (cost=8.86..27.62 rows=5 width=69) (actual time=0.039..0.046 rows=4 loops=1)
                           Recheck Cond: ((name)::text = ANY ('{rnv,fop}'::text[]))
                           Heap Blocks: exact=4
                           ->  Bitmap Index Scan on ix_package_name  (cost=0.00..8.86 rows=5 width=0) (actual time=0.033..0.033 rows=4 loops=1)
                                 Index Cond: ((name)::text = ANY ('{rnv,fop}'::text[]))
               ->  Index Scan using build_pkey on build build_1  (cost=0.42..8.44 rows=1 width=55) (actual time=0.007..0.007 rows=1 loops=2)
                     Index Cond: (id = package.last_complete_build_id)
         ->  Materialize  (cost=0.00..1.04 rows=3 width=311) (actual time=0.005..0.007 rows=3 loops=2)
               ->  Seq Scan on collection collection_1  (cost=0.00..1.03 rows=3 width=311) (actual time=0.002..0.004 rows=3 loops=1)
 Planning time: 0.726 ms
 Execution time: 0.326 ms
(24 rows)
```